src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -0,0 +1,617 @@
+    rearRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+  }
+
+  public Pigeon2 getGyro() {
+    return gyro;
+  }
+
+  /** adapted from last year's version, which was taken from Github */
+  public Command followTrajectoryCommand(PathPlannerPath path, boolean isFirstPath) {
+    return new SequentialCommandGroup(
+        new InstantCommand(
+            () -> {
+              // Reset odometry for the first path you run during auto
+              if (isFirstPath) {
+                this.resetOdometry(path.getPreviewStartingHolonomicPose());
+              }
+            }),
+
+        new FollowPathHolonomic( // old method from last year: PPSwerveControllerCommand
+            path,
+            this::getPose, // pose supplier
+            this::getChassisSpeeds, // robot relative chassis speeds supplier
+            this::setOutputRobotRelativeSpeeds, // robot relative chassis speeds consumer
+            new PIDConstants(DriveCal.DRIVING_P, DriveCal.DRIVING_I, DriveCal.DRIVING_D), // PID Constants for
+                                                                                          // translation
+            new PIDConstants(DriveCal.TURNING_P, DriveCal.TURNING_I, DriveCal.TURNING_D), // PID Constants for rotation
+            DriveConstants.DRIVE_WHEEL_FREE_SPEED_METERS_PER_SECOND, // maxModuleSpeed
+            DriveConstants.DRIVE_BASE_RADIUS_METERS,
+            new ReplanningConfig(), // creates a path replanning configuration with the default config
+            () -> {
+              return false;
+            }, // boolean supplier for shouldFlipPath
+            this),
+
+        new InstantCommand(
+            () -> {
+              targetHeadingDegrees = getHeadingDegrees();
+            }),
+
+        new PrintCommand("Finished a trajectory"));
+  }
+
+  /** Returns a past pose, given a latency adjustment */
+  public Pose2d getPastPose(double latencySec) {
+    Pose2d curPose = getPose();
+    double latencyAdjustmentSec = 0.00;
+    latencySec += latencyAdjustmentSec;
+    Transform2d pastTransform = new Transform2d(
+        new Translation2d(
+            -lastSetChassisSpeeds.vxMetersPerSecond * latencySec,
+            -lastSetChassisSpeeds.vyMetersPerSecond * latencySec),
+        Rotation2d.fromRadians(lastSetChassisSpeeds.omegaRadiansPerSecond * latencySec)
+            .unaryMinus());
+    Pose2d pastPose = curPose.plus(pastTransform);
+    return pastPose;
+  }
+
+  public void setLimelightTargetFromTransform(Transform2d transform, double latencySec) {
+    // Transform is to get the limelight to the correct location, not to get the
+    // robot
+    // Here we correct for that
+    Transform2d flipTransform = new Transform2d(
+        new Translation2d(-transform.getX(), transform.getY()), transform.getRotation());
+
+    double latencyAdjustmentSec = 0.00;
+
+    Pose2d curPose = getPose();
+    Pose2d pastPose = getPastPose(latencyAdjustmentSec);
```

Comment:
Let's add a todo to see if we can get this working with the real latency `latencySec` since right now the latency compensation is basically just turned off